lib/bap_image/llvm_binary_stubs.cpp

```diff
@@ -0,0 +1,172 @@
+#include <memory>
+#include <string>
+
+#include "llvm_binary_stubs.h"
+#include "llvm_binary_stubs.hpp"
+
+#include <llvm/Support/MemoryBuffer.h>
+#include <llvm/Object/Binary.h>
+#include <llvm/Support/system_error.h>
+#include <llvm/ADT/Triple.h>
+#include <llvm/Object/Archive.h>
+#include <llvm/Object/ObjectFile.h>
+#include <llvm/Support/Casting.h>
+
+#include "llvm_binary_stubs.h"
+#include "llvm_segment.hpp"
+
+extern "C" {
+#include <caml/mlvalues.h>
+#include <caml/memory.h>
+#include <caml/fail.h>
+#include <caml/alloc.h>
+#include <caml/custom.h>
+#include <caml/bigarray.h>
+}
+
+void llvm_binary_fail [[ noreturn ]](const char* message) {
+    ::caml_failwith(message);
+}
+
+namespace impl {
+
+using namespace llvm;
+using namespace llvm::object;
+
+Binary* from_value (::value v) {
+    return *reinterpret_cast<Binary**>(Data_custom_val(v));
+}
+
+::value to_value (Binary* b) {
+    static struct custom_operations binary_ops = {
+        const_cast<char*>("llvm.caml.llvm_binary"),
+        [](::value v) { delete from_value(v); },
+        custom_compare_default,
+        custom_hash_default,
+        custom_serialize_default,
+        custom_deserialize_default
+    };
+    ::value v = ::alloc_custom(&binary_ops, sizeof(Binary*), 0, 1);
+    *reinterpret_cast<Binary**>(Data_custom_val(v)) = b;
+    return v;
+}
+
+
+::value to_value(Triple::ArchType arch) {
+    typedef Triple::ArchType AT;
+    switch(arch) {
```

Comment:
It is better to use strings here. For example, what if some of the variants has clashing hashes?